networkx/algorithms/tree/distance_measures.py

```diff
@@ -0,0 +1,129 @@
+    ----------
+    G : NetworkX graph
+       A tree.
+
+    root : node
+       A node in `G`.
+
+    Returns
+    -------
+    s : dict
+       Dictionary of subtree sizes keyed on nodes.
+
+    Examples
+    --------
+    >>> _subtree_sizes(nx.path_graph(4), 0)
+    {0: 4, 1: 3, 2: 2, 3: 1}
+
+    >>> _subtree_sizes(nx.path_graph(4), 2)
+    {2: 4, 1: 2, 0: 1, 3: 1}
+    """
+    sizes = {root: 1}
+    stack = [root]
+    for a, b in nx.dfs_edges(G, root):
+        while stack[-1] != a:
+            x = stack.pop()
+            sizes[stack[-1]] += sizes[x]
+        stack.append(b)
+        sizes[b] = 1
+    for b, a in nx.utils.pairwise(reversed(stack)):
+        sizes[a] += sizes[b]
+    return sizes
+
+
+@nx.utils.not_implemented_for("directed")
+@nx._dispatchable
+def tree_centroid(G):
+    """Return the centroid of an unweighted tree.
+
+    The centroid of a tree is the set of nodes such that removing any
+    one of them would split the tree into a forest of subtrees, each
+    with at most ``N / 2`` nodes, where ``N`` is the number of nodes
+    in the original tree. This set may contain two nodes if removing
+    an edge between them results in two trees of size exactly ``N /
+    2``.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+       A tree.
+
+    Returns
+    -------
+    c : list
+       List of nodes in centroid of the tree. This could be one or two nodes.
+
+    Raises
+    ------
+    NotATree
+        If the input graph is not a tree.
+    NotImplementedException
+        If the input graph is directed.
+    NetworkXPointlessConcept
+        If `G` has no nodes or edges.
+
+    Notes
+    -----
+    This algorithm's time complexity is ``O(N)`` where ``N`` is the
+    number of nodes in the tree.
+
+    In unweighted trees the centroid coincides with the barycenter,
+    the node or nodes that minimize the sum of distances to all other
+    nodes.  However, this concept is different from that of the graph
+    center, which is the set of nodes minimizing the maximum distance
+    to all other nodes.
+
+    Examples
+    --------
+    >>> G = nx.path_graph(4)
+    >>> nx.tree_centroid(G)
+    [1, 2]
+
+    A star-shaped tree with one long branch illustrates the difference
+    between the centroid and the center. The center lies near the
+    middle of the long branch, minimizing maximum distance. The
+    centroid, however, limits the size of any resulting subtree to at
+    most half the total nodes, forcing it to remain near the hub when
+    enough short branches are present.
+
+    >>> G = nx.star_graph(6)
+    >>> nx.add_path(G, [6, 7, 8, 9, 10])
+    >>> nx.tree_centroid(G), nx.center(G)
+    ([0], [7])
+
+    See Also
+    --------
+    :func:`~networkx.algorithms.distance_measures.barycenter`
+    :func:`~networkx.algorithms.distance_measures.center`
+
+    """
+    if not nx.is_tree(G):
+        raise nx.NotATree("provided graph is not a tree")
+    prev, root = None, next(iter(G.nodes))
```

Comment:
In the rest of the codebase this is usually done like this
```suggestion
    prev, root = None, nx.utils.arbitrary_element(G)
```